### PR TITLE
Fix treatment of bottom types in OrType#join and baseType computations

### LINKED ISSUE #11968: Unable to select member of nullable user-defined class
```
scalac Stest.scala
```

```Scala
class C {
    def get(): Int = 0
}

def g = {
    val s: String | Null = ???
    val l = s.length // ok
    val c: C | Null = ???
    c.get() // error: value get is not a member of C | Null
}
```

## Output

```
-- [E008] Not Found Error: Stest.scala:13:6 ------------------------------------
13 |    c.get()
   |    ^^^^^
   |    value get is not a member of C | Null
```

The nullable `C` is expected to have similar behaviour as nullable `String`.

I think the issue is at `TypeOps.orDominator`. `String | Null` after join becomes `String`. But `C | Null` after join becomes `Object`.

This code can be compiled using explicit-nulls and unsafeNulls (`scalac -Yexplicit-nulls -language:unsafeNulls`), since unsafeNulls will ignore Null type when searching members, then there is no need to join the OrType (see `Types.goOr`).

 `(String | Null).join` can return a correct result, because the type of `String` after `widenExpr` is different, `approximateOr` uses subtyping and `lub` to decide the result. 

```scala
// approximateOr 
if ((tp1w ne tp1) || (tp2w ne tp2)) {
  val isSingle1 = tp1.isInstanceOf[SingletonType]
  val isSingle2 = tp2.isInstanceOf[SingletonType]
  return {
    if (tp2w eq tp2) orDominator(tp1w | tp2)                  // 2.1
    else if (tp1w eq tp1) orDominator(tp1 | tp2w)             // 2.1
    else if (tp1w frozen_<:< tp2w) orDominator(tp1w | tp2)    // 2.2
    else if (tp2w frozen_<:< tp1w) orDominator(tp1 | tp2w)    // 2.2
    else if (isSingle1 && !isSingle2) orDominator(tp1w | tp2) // 2.3
    else if (isSingle2 && !isSingle1) orDominator(tp1 | tp2w) // 2.3
    else if (tp1 frozen_<:< tp2w) tp2w                        // 2.4
    else if (tp2 frozen_<:< tp1w) tp1w                        // 2.5
    else orDominator(tp1w | tp2)                              // 2.6
  }
}
```

However, the type of `C` after `widenExpr` is the same, `approximateOr` uses `baseClasses` of `C` and `Null` to decide. I think we need to handle bottom classes for commonBaseClasses more carefully to fix the issue.

```scala
// Step 3: Intersect base classes of both sides
val commonBaseClasses = tp.mapReduceOr(_.baseClasses)(intersect)
val doms = dominators(commonBaseClasses, Nil)
def baseTp(cls: ClassSymbol): Type =
  tp.baseType(cls).mapReduceOr(identity)(mergeRefinedOrApplied)
doms.map(baseTp).reduceLeft(AndType.apply)
```

I don't understand the reason to use baseClasses here. Why don't we using subtyping compare in step 3 as well?

### PR DESCRIPTION
**Pull request by odersky**
Fixes #11968

### CODE REVIEW

**Comment by noti0na1** [comment]
Could you extend test to this? Thanks

```scala
// Test selecting member of nullable classes

class A {
  def get(): Int = 0
}

class B extends A {}

class C extends A {}

def test1 = {
  val s: String | Null = ???
  val l = s.length

  val a: A | Null = new A
  a.get()

  val bc: B | C = new B
  bc.get()

  val bcn: B | (C | Null) = new C
  bcn.get()

  val bnc: (B | Null) | C = null
  bnc.get()

  val abcn: A | B | C | Null = new A
  abcn.get()
}

def test2 = {
  val s: String | Nothing = ???
  val l = s.length

  val a: A | Nothing = new A
  a.get()

  val bc: B | C = new B
  bc.get()

  val bcn: B | (C | Nothing) = new C
  bcn.get()

  val bnc: (B | Nothing) | C = new B
  bnc.get()

  val abcn: A | B | C | Nothing = new A
  abcn.get()
}
```

File: compiler/src/dotty/tools/dotc/core/TypeOps.scala [hunk 1]
```diff
@@ -194,12 +194,16 @@ object TypeOps:
...
-    def intersect(cs1: List[ClassSymbol], cs2: List[ClassSymbol]): List[ClassSymbol] = {
-      val cs2AsSet = new util.HashSet[ClassSymbol](128)
-      cs2.foreach(cs2AsSet += _)
-      cs1.filter(cs2AsSet.contains)
-    }
+    /** a faster version of cs1 intersect cs2 that treats bottom types correctly */
+    def intersect(cs1: List[ClassSymbol], cs2: List[ClassSymbol]): List[ClassSymbol] =
+      if cs1.head == defn.NothingClass then cs2
+      else if cs2.head == defn.NothingClass then cs1
+      else if cs1.head == defn.NullClass && !ctx.explicitNulls && cs2.head.derivesFrom(defn.ObjectClass) then cs2
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/TypeOps.scala:201]: I don't think we need `derivesFrom` here. Can we just look up `ObjectClass` in cs2? `cs2.contains(defn.ObjectClass)`

File: tests/pos/i11968.scala [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class C {
+    def get(): Int = 0
+}
+
+def g = {
+    val s: String | Null = ???
+    val l = s.length // ok
+    val c: C | Null = ???
+    c.get() // error: value get is not a member of C | Null
```
Comment (noti0na1) [inline tests/pos/i11968.scala:9]: Could you remove the comment here?

File: compiler/src/dotty/tools/dotc/core/TypeOps.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeOps.scala:201]: Looking it up is actually less efficient.

File: tests/pos/i11968.scala (hunk 2 above)
Comment (odersky) [inline tests/pos/i11968.scala:9]:
```suggestion
    c.get() 
```

**Review by noti0na1 — APPROVED** [review]
LGTM

**Pull request merged**